# Лабораторная работа № 1 Линейная регрессия

## Задание
Решите пожалуйста следующие задачи.

1. Примените следующие типы регрессий для датасета представленного по
[ссылке](https://drive.google.com/file/d/1LBDnhITL0Wqwp5G6M6IBI-SSz8BIoNec/view?usp=share_link).
- Линейная регрессия
- LASSO регрессия
- Ridge регрессия
- ElasticNet регрессия
- Polynom регрессия (для степени полинома 2)
- Polynom регрессия (для степени полинома 5)
- Polynom регрессия (для степени полинома 10)

Постройте модель для атрибутов `X = {SkinThickness, BMI}` и `Y = {Insulin}`.  
**Укажите `score` моделей.**


## Загрузка данных

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
! gdown --fuzzy 'https://drive.google.com/file/d/1LBDnhITL0Wqwp5G6M6IBI-SSz8BIoNec/view?usp=share_link'

Downloading...
From: https://drive.google.com/uc?id=1LBDnhITL0Wqwp5G6M6IBI-SSz8BIoNec
To: /content/diabetes.csv
100% 23.1k/23.1k [00:00<00:00, 53.6MB/s]


In [ ]:
df = pd.read_csv('/content/diabetes.csv')
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

##  О датасете
[Pima Indians Diabetes Database](https://www.kaggle.com/datasets/uciml/pima-indians-diabetes-database/data)

Этот набор данных взят из Национального института диабета, заболеваний органов пищеварения и почек (NIDDK). Целью набора данных является диагностическое предсказание наличия у пациента диабета на основе определенных диагностических измерений, включенных в набор данных. На выбор этих экземпляров из более крупной базы данных было наложено несколько ограничений. В частности, все пациенты здесь — женщины не моложе 21 года, родом из индейцев Пима.

Наборы данных состоят из нескольких переменных медицинских предикторов и одной целевой переменной `Outcome`. Переменные-предикторы включают количество беременностей, которые были у пациентки, её индекс массы тела `BMI`, уровень инсулина, возраст и так далее. Всего в датасете 9 переменных:
- `Pregnancies` - количество беременностей;
- `Glucose` - концентрация глюкозы в плазме через 2 часа при пероральном тесте на толерантность к глюкозе;
- `BloodPressure` - диастолическое артериальное давление (мм рт. ст.);
- `SkinThickness` - толщина кожной складки трицепса (мм);
- `Insulin` - 2-часовой сывороточный инсулин (мкЕд/мл);
- `BMI` - Индекс массы тела (вес в кг/(рост в м)^2);
- `DiabetesPedigreeFunction` - функция оценки вероятности диабета на основе семейного анамнеза;
- `Age` - возраст (лет);
- `Outcome` - целевая переменная (0 или 1) 268 из 768 — 1 (больны), остальные — 0 (здоровы).



In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [ ]:
df.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

In [ ]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

## Линейная регрессия

In [ ]:
# подключение библиотек
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
# Сформируем матрицу объект-признак X и вектор целевой переменной y
X = df[['SkinThickness', 'BMI']]
y = df['Insulin']

X.head()

SkinThickness   BMI
0             35  33.6
1             29  26.6
2              0  23.3
3             23  28.1
4             35  43.1

In [ ]:
# Разобъем данные на обучающую и тестовую части
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

X_train.shape, X_test.shape

((614, 2), (154, 2))

In [ ]:
# Сформируем и обучим линейную регрессию
linear_reg = LinearRegression()
linear_reg.fit(X_train, y_train)

LinearRegression()

In [ ]:
# Сделаем предсказание на основе тестовых значений
y_pred_linear = linear_reg.predict(X_test)

In [ ]:
# оценим модель
score_linear = r2_score(y_test, y_pred_linear)
print(f"Linear Regression R^2 Score: {score_linear}")

Linear Regression R^2 Score: 0.18117151678040866


## LASSO регрессия

In [ ]:
# подключение библиотеки Lasso
from sklearn.linear_model import Lasso

# инициализируем и обучаем Lasso регрессию
lasso_reg = Lasso(alpha=0.1) # alpha — величина регуляризации
lasso_reg.fit(X_train, y_train)

# предсказание результата
y_pred_lasso = lasso_reg.predict(X_test)

# оценка модели
score_lasso = r2_score(y_test, y_pred_lasso)

print(f"LASSO Regression R^2 Score: {score_lasso}")

LASSO Regression R^2 Score: 0.18118401492198166


## Ridge регрессия

In [ ]:
# подключение библиотеки Ridge
from sklearn.linear_model import Ridge

# инициализируем и обучаем ridge регрессию
ridge_reg = Ridge(alpha=1.0)
ridge_reg.fit(X_train, y_train)

# предсказание результата
y_pred_ridge = ridge_reg.predict(X_test)

# оценка модели
score_ridge = r2_score(y_test, y_pred_ridge)

print(f"Ridge Regression R^2 Score: {score_ridge}")

Ridge Regression R^2 Score: 0.18117215005550003


## ElasticNet регрессия

In [ ]:
# подключение библиотеки ElasticNet
from sklearn.linear_model import ElasticNet

# иниализируем и обучаем ElasticNet регрессию ци
elastic_net_reg = ElasticNet(alpha=1.0, l1_ratio=0.5)
elastic_net_reg.fit(X_train, y_train)

# предсказание результата
y_pred_elastic_net = elastic_net_reg.predict(X_test)

# оценка модели
score_elastic_net = r2_score(y_test, y_pred_elastic_net)

print(f"ElasticNet Regression R^2 Score: {score_elastic_net}")

ElasticNet Regression R^2 Score: 0.1814302410119406


## Polynom регрессия (для степени полинома 2)

In [ ]:
# подключение библиотеки PolynomialFeatures
from sklearn.preprocessing import PolynomialFeatures

# генерация полиномиальных признаков
poly_features_2 = PolynomialFeatures(degree=2)
X_train_poly_2 = poly_features_2.fit_transform(X_train)
X_test_poly_2 = poly_features_2.transform(X_test)

# инициализируем и обучаем PolynomialFeatures регрессию
poly_reg_2 = LinearRegression()
poly_reg_2.fit(X_train_poly_2, y_train)

# предсказание результата
y_pred_poly_2 = poly_reg_2.predict(X_test_poly_2)

# оценка модели
score_poly_2 = r2_score(y_test, y_pred_poly_2)

print(f"Polynomial Regression (degree 2) R^2 Score: {score_poly_2}")

Polynomial Regression (degree 2) R^2 Score: 0.16803904638654665


## Polynom регрессия (для степени полинома 5)

In [ ]:
# подключение библиотеки PolynomialFeatures
from sklearn.preprocessing import PolynomialFeatures

# генерация полиномиальных признаков
poly_features_5 = PolynomialFeatures(degree=5)
X_train_poly_5 = poly_features_5.fit_transform(X_train)
X_test_poly_5 = poly_features_5.transform(X_test)

# инициализируем и обучаем PolynomialFeatures регрессию
poly_reg_5 = LinearRegression()
poly_reg_5.fit(X_train_poly_5, y_train)

# предсказание результата
y_pred_poly_5 = poly_reg_5.predict(X_test_poly_5)

# оценка модели
score_poly_5 = r2_score(y_test, y_pred_poly_5)

print(f"Polynomial Regression (degree 5) R^2 Score: {score_poly_5}")

Polynomial Regression (degree 5) R^2 Score: -102.11290812715477


## Polynom регрессия (для степени полинома 10)

In [ ]:
# подключение библиотеки PolynomialFeatures
from sklearn.preprocessing import PolynomialFeatures

# генерация полиномиальных признаков
poly_features_10 = PolynomialFeatures(degree=10)
X_train_poly_10 = poly_features_10.fit_transform(X_train)
X_test_poly_10 = poly_features_10.transform(X_test)

# инициализируем и обучаем PolynomialFeatures регрессию
poly_reg_10 = LinearRegression()
poly_reg_10.fit(X_train_poly_10, y_train)

# предсказание результата
y_pred_poly_10 = poly_reg_10.predict(X_test_poly_10)

# оценка модели
score_poly_10 = r2_score(y_test, y_pred_poly_10)

print(f"Polynomial Regression (degree 10) R^2 Score: {score_poly_10}")

Polynomial Regression (degree 10) R^2 Score: -30000512581.10792


## Вывод

Таким образом, можно сделать вывод о том, что результаты оценок моделей линейной регрессии, LASSO регрессии, Ridge регрессии и ElasticNet регрессии очень схожи и равны приближённо 0,18.

Значения, полученные при оценке с помощью Polynom регрессии 2, 5 и 10 степеней, имеют большие отличия, так:
* Polynom регрессия степени 2 имеет значение 0,168, то есть меньшее, чем те, которые были предсказаны с помощью линейной регрессии, LASSO регрессии, Ridge регрессии и ElasticNet регрессии.
*  Polynom регрессия степени 2 и 10 имеют очень низкие отрицательные значения оценок. Значит при добавлении полиномиальных признаков  модель работает хуже.

В целом можно сказать, что наилучшее значение показала ElasticNet регрессия.
Но все-равно значение качества модели очень низкое.

Это может говорить о том, что связь между целевой переменной Insulin и признаками SkinThickness и BMI - незначительная. Для получения лучших результатов следует рассмотреть связь с другими признаками.